extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java

```diff
@@ -135,7 +137,39 @@ protected Object getCacheKey(Cache cache, List<Short> cacheKeyParameterPositions
         }
     }
 
-    protected static boolean isUniReturnType(InvocationContext invocationContext) {
-        return Uni.class.isAssignableFrom(invocationContext.getMethod().getReturnType());
+    protected static ReturnType determineReturnType(Class<?> returnType) {
+        if (Uni.class.isAssignableFrom(returnType)) {
+            return ReturnType.Uni;
+        }
+        if (CompletionStage.class.isAssignableFrom(returnType)) {
+            return ReturnType.CompletionStage;
+        }
+        return ReturnType.NonAsync;
+    }
+
+    protected Uni<?> asyncInvocationResultToUni(Object invocationResult, ReturnType returnType) {
+        if (returnType == ReturnType.Uni) {
+            return (Uni<?>) invocationResult;
+        } else if (returnType == ReturnType.CompletionStage) {
+            return Uni.createFrom().completionStage((CompletionStage<?>) invocationResult);
```

Comment:
I'm not sure `@CacheResult#lockTimeout` will work if the `Uni` is created from the `CompletionStage` here. Delaying the `CompletionStage` computation (with a `Supplier<CompletionStage>`) until the `Uni` is resolved may be necessary:
```suggestion
            return Uni.createFrom().completionStage(() -> (CompletionStage<?>) invocationContext.proceed());
```
(the real code would need a proper `Supplier` and not that lambda 😄)